### PR TITLE
chore(agw): Python dependency upgrades

### LINKED ISSUE #13426: Upgrade Python dependencies after python 3.8 update
## Subtasks

- [x] Try the branch from sebastians pr https://github.com/magma/magma/pull/12972
- [x] Test prod package creation locally by setting up a magma vm. Building the python packages (ask nils for the command).  Create magma debian package. Install magma debian package and dependencies.
- [ ] Inform 5G test folks that magma debian packages have been touched
- [x] Run integration tests with updated dependencies (lte/s1ap)
- [ ] Make sure that build all job runs for two consecutive runs on master

## Description

Many Python dependencies have not been updated in a while. But now that we have Python 3.8 everywhere a lot of these upgrades could be made.

An attempt was made here:

https://github.com/magma/magma/pull/12972

There are multiple steps necessary that make this a little complicated:

- Increase versions in setup.py files (as in the PR above). As a first step, try to avoid incompatible upgrades that make significant code changes necessary. There are a lot of low-hanging fruits, i.e. upgrades that just work.
- Also increase versions in requirements.in to make bazel happy (possibly need to generate requirements.txt afterwards?)
- Test that things still work. Unit, sudo and integration tests of VM.
- Create a new production release with the new versions. Make sure the production versions are also upgraded (depends on the whims of pydep).
- Test the new production release can actually be installed and works correctly. Use the new production VM integration test tool from https://github.com/magma/magma/issues/13263.
 

Hint: Lucas Gonze was also looking into some of these upgrades, so at least some of our existing versions have security issues.
https://github.com/magma/magma/pull/13222

### PR DESCRIPTION
**Pull request by MoritzThomasHuebner**
Fixes #13426

In pairing with @wolfseb 

## Summary

Done:
- Upgraded Python versions where possible. We used `pipdeptree` to analyze packages and dependencies.
- Temporary change to upgrade `fpm` during provisioning. This is necessary to build `cryptography`.
- Some packages (such as docker) are not built by `pydep`, since a `python3-docker` exists in `http://archive.ubuntu.com/ubuntu/`.
- Updated lockfile

## Test Plan

-  Run `fab release package:destroy_vm=True` on the host VM to trigger new magma and dependency `*.deb` files to be built.
- Install local `*.deb` files, and any other dependencies that have not been newly packaged from the artifactory.
- Remove `/etc/magma`.
- Install `magma*.deb` file.
- Spin up `magma_test` and `magma_trfserver` 
- Run S1AP precommit and integration tests.
- Run `build-all` script on a fork: https://github.com/MoritzThomasHuebner/magma/runs/7759753398?check_suite_focus=true
  - For this run we needed to make a temporary change that disables the check that we are on the origin repo.
  - The check shows that all packages are actually being built.
- Run Bazel build and test workflow on fork: https://github.com/MoritzThomasHuebner/magma/actions/runs/2830383274
  - Ensures versions in `bazel/external/requirements.in` do not cause any version conflicts

## Additional Information

There are a number of packages we were not able to upgrade:
- `fpm` requires that there is a `setup.py`, thus some of the newest versions of `pyparsing`, `jsonschema`, `pyroute2` can no longer be built since these packages now utilize other build methods.
- `setuptools`: The version `49.6.0` is hardcoded in some places.
- `docker`: Ubuntu focal is not compatible with `docker>=5`.
  - `requests`: The latest versions require `docker>=5`.
- `flask` is restricted to `1.1.4`.
  - This restricts `click`, `itsdangerous`, `Jinja2`, and `markupsafe`.
- `redis` is restricted to `3.5.3`.
- `cryptography` is restricted to `3.2.1` because newer versions cannot be built with `fpm`.
- `PyYAML` is restricted to `<6.0`

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by alexzurbonsen** [comment]
For documentation purposes: running `fab release package:destroy_vm=True` creates the following python3 packages

```
[VM] ~/magma-packages $ ls -1 | grep python3
python3-aiodns_3.0.0_all.deb
python3-aiohttp_3.8.1_amd64.deb
python3-aiosignal_1.2.0_all.deb
python3-async-timeout_4.0.2_all.deb
python3-bravado-core_5.17.0_all.deb
python3-certifi_2022.6.15_all.deb
python3-cffi_1.14.5_amd64.deb
python3-charset-normalizer_2.0.0_all.deb
python3-click_8.1.3_all.deb
python3-cryptography_3.2.1_amd64.deb
python3-cython_0.29.32_amd64.deb
python3-fire_0.4.0_all.deb
python3-freezegun_1.2.1_all.deb
python3-frozenlist_1.3.1_amd64.deb
python3-grpcio_1.47.0_amd64.deb
python3-idna_3.3_all.deb
python3-itsdangerous_2.1.2_all.deb
python3-jinja2_3.1.2_all.deb
python3-json-pointer_2.3_all.deb
python3-jsonpickle_2.2.0_all.deb
python3-lxml_4.9.1_amd64.deb
python3-markupsafe_2.1.1_amd64.deb
python3-multidict_6.0.2_amd64.deb
python3-netifaces_0.11.0_amd64.deb
python3-psutil_5.9.1_amd64.deb
python3-pycares_4.2.1_amd64.deb
python3-pycparser_2.20_all.deb
python3-pycryptodome_3.15.0_amd64.deb
python3-pyroute2.core_0.6.13_all.deb
python3-pyroute2.ethtool_0.6.13_all.deb
python3-pyroute2.ipdb_0.6.13_all.deb
python3-pyroute2.ipset_0.6.13_all.deb
python3-pyroute2.ndb_0.6.13_all.deb
python3-pyroute2.nftables_0.6.13_all.deb
python3-pyroute2.nslink_0.6.13_all.deb
python3-pyroute2_0.6.13_all.deb
python3-python-dateutil_2.8.2_all.deb
python3-redis-collections_0.11.0_all.deb
python3-requests_2.28.1_all.deb
python3-scapy_2.4.5_all.deb
python3-sentry-sdk_1.9.0_all.deb
python3-six_1.16.0_all.deb
python3-spyne_2.14.0_all.deb
python3-urllib3_1.25.3_all.deb
python3-webcolors_1.12_all.deb
python3-yaml_6.0_amd64.deb
python3-yarl_1.8.1_amd64.deb
```

**Comment by alexzurbonsen** [comment]
Tried to install the magma package and python3 dependencies on a regular VM per @nstng instructions (as in the PR description):
- delete `/etc/magma`
- apt install python3 dependencies and magma*.deb package
- run a set of commands that copies systemd service files, does some linking etc (not yet committed)

Afterwards services where not running properly since packages could not be found. Had to reinstall some packages (e.g. in magmad `sudo apt --reinstall install python3-h2`). In that way I got magmad running. 

Next, I tried pipelined. Turns out that we seem to use a jinja2 version that is too high.* Instead of 3.1.2 we should probably try the highest v2, that is 2.11.3. 

It might also be a better idea, to use the prod VM for installation and testing. May be, there we don't face all the  issues with reinstalling packages. 

\* `ImportError: cannot import name 'escape' from 'jinja2' (/usr/local/lib/python3.8/dist-packages/jinja2/__init__.py)`

**Comment by MoritzThomasHuebner** [comment]
> Tried to install the magma package and python3 dependencies on a regular VM per @nstng instructions (as in the PR description):
> 
>     * delete `/etc/magma`
> 
>     * apt install python3 dependencies and magma*.deb package
> 
>     * run a set of commands that copies systemd service files, does some linking etc (not yet committed)
> 
> 
> Afterwards services where not running properly since packages could not be found. Had to reinstall some packages (e.g. in magmad `sudo apt --reinstall install python3-h2`). In that way I got magmad running.
> 
> Next, I tried pipelined. Turns out that we seem to use a jinja2 version that is too high.[1] Instead of 3.1.2 we should probably try the highest v2, that is 2.11.3.
> 
> It might also be a better idea, to use the prod VM for installation and testing. May be, there we don't face all the issues with reinstalling packages.
> 
> [1] `ImportError: cannot import name 'escape' from 'jinja2' (/usr/local/lib/python3.8/dist-packages/jinja2/__init__.py)`

Thanks for the comment @alexzurbonsen, @wolfseb and I have roughly the same situation. It is still not entirely clear to us why some of the services are not running. Packages like `h2` should have been installed from artifactory with the installation of the python3 dependencies.

**Comment by wolfseb** [comment]
Python packages that need to be reinstalled via 
```
sudo apt reinstall python3-*
```
* prometheus-client
* h2
* jinja2
* eventlet
* redis
* ovs
* envoy

Then all magma services should be running (restart via systemctl may be necessary)

**Comment by wolfseb** [comment]
Integ tests have been run for the new VM. The following tests fail:

Precommit:


Extended:
tba

Nonsanity:
- test_activate_deactivate_multiple_dedicated.py
- test_sctp_shutdown_while_mme_is_stopped.py
- test_3495_timer_for_dedicated_bearer_with_mme_restart.py
- test_no_attach_complete_with_mme_restart.py

File: .github/workflows/bazel.yml
Comment (sebathomas) [inline .github/workflows/bazel.yml]: I think these commented-out filters need to be reverted?

File: .github/workflows/build_all.yml
Comment (sebathomas) [inline .github/workflows/build_all.yml]: Same here.

File: lte/gateway/deploy/roles/dev_common/tasks/main.yml
Comment (sebathomas) [inline lte/gateway/deploy/roles/dev_common/tasks/main.yml]: Nit: Usually, we just set the step to "full_provision" with a TODO rather than duplicating it.

**Review by sebathomas — CHANGES_REQUESTED** [review]
This is looking good, except for the commented-out code. Thanks for finally getting these updates working!

By the way, I think "chore(agw)" would be a more fitting tag than "ci".

File: .github/workflows/bazel.yml
Comment (MoritzThomasHuebner) [reply .github/workflows/bazel.yml]: Done, we commented these out temporarily to run the workflow on my fork.

File: .github/workflows/build_all.yml
Comment (MoritzThomasHuebner) [reply .github/workflows/build_all.yml]: Done

File: lte/gateway/deploy/roles/dev_common/tasks/main.yml
Comment (MoritzThomasHuebner) [reply lte/gateway/deploy/roles/dev_common/tasks/main.yml]: Done

**Review by sebathomas — APPROVED** [review]

**Comment by sebathomas** [comment]
Needs a rebase and reviews from @magma/approvers-bazel and @magma/approvers-agw-pipelined .

**Review by nstng — APPROVED** [review]
bazel + pipelined lgtm

**Comment by nstng** [comment]
> Needs a rebase and reviews from @magma/approvers-bazel and @magma/approvers-agw-pipelined .

FYI: do not try to merge the requirements.txt - please rebuild it after the rebase

**Comment by MoritzThomasHuebner** [comment]
> > Needs a rebase and reviews from @magma/approvers-bazel and @magma/approvers-agw-pipelined .
> 
> FYI: do not try to merge the requirements.txt - please rebuild it after the rebase

@nstng thanks for letting me know. I'm a bit confused, though. There is now `docker>=4.0.2` in the `requirements.in`, which adds `docker==5.0.3` in the `requirements.txt`. However, we found that we can't use `docker==5.x` on ubuntu focal since the latest version on http://archive.ubuntu.com/ubuntu is [4.1.0](https://packages.ubuntu.com/focal/python3-docker). While I think you can install later versions manually with `pip`, the workflow will always default to what is the latest version online if it exists.

**Comment by nstng** [comment]
> @nstng thanks for letting me know. I'm a bit confused, though. There is now `docker>=4.0.2` in the `requirements.in`, which adds `docker==5.0.3` in the `requirements.txt`. However, we found that we can't use `docker==5.x` on ubuntu focal since the latest version on http://archive.ubuntu.com/ubuntu is [4.1.0](https://packages.ubuntu.com/focal/python3-docker). While I think you can install later versions manually with `pip`, the workflow will always default to what is the latest version online if it exists.

I added `docker>=4.0.2` for `orc8r/gateway/python/magma/common/health/docker_health_service.py` and this is actually the only python docker import I see in magma. In `orc8r/gateway/python/setup.py` we have `docker>=4.0.2` (sure, the stricter restriction will be used for `pydep`). I did not found a problem with a more recent docker version - did you? I assume that the `lte` `setup.py` is outdated.
I am unsure what you mean by
> the workflow will always default to what is the latest version online if it exists.

**Review by nstng — CHANGES_REQUESTED** [review]
Will block this intentionally because we have a version issue with `sentry-sdk` in the 1.8 RC that we might want to fix less invasive and merge first for backporting.
See https://github.com/magma/magma/issues/13645

**Comment by sebathomas** [comment]
Bazel seems to have an issue with the systemd dependency after the rebase: https://github.com/magma/magma/runs/7892895935

**Comment by alexzurbonsen** [comment]
fyi: This PR includes some of the work required by https://github.com/magma/magma/issues/11743.

**Comment by sebathomas** [comment]
I think the CI error in this PR is because the Bazel requirements.txt was not regenerated after the latest rebase. (Maybe it would be good to have this as part of the in-sync check?)

When I try it on my machine, I get some errors though:
```
Could not find a version that matches markupsafe==1.1.1,>=2.0 from https://files.pythonhosted.org/packages/b9/2e/64db92e53b86efccfaea71321f597fa2e1b2bd3853d8ce658568f7a13094/MarkupSafe-1.1.1.tar.gz (from -r requirements.in (line 22))
Tried: 0.9, 0.9.1, 0.9.2, 0.9.3, 0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.20, 0.21, 0.22, 0.23, 1.0, 1.1.0, 1.1.1, 2.0.0, 2.0.1, 2.0.1, 2.1.0, 2.1.0, 2.1.1, 2.1.1
Skipped pre-versions: 2.0.0a1, 2.0.0rc1, 2.0.0rc2
There are incompatible versions in the resolved dependencies:
  markupsafe==1.1.1 from https://files.pythonhosted.org/packages/b9/2e/64db92e53b86efccfaea71321f597fa2e1b2bd3853d8ce658568f7a13094/MarkupSafe-1.1.1.tar.gz (from -r requirements.in (line 22))
  MarkupSafe>=2.0 (from jinja2==3.1.2->-r requirements.in (line 52))
```

**Comment by sebathomas** [comment]
I think I fixed it. Used the same versions in the new scripts requirements as in the setup.py.

**Comment by sebathomas** [comment]
Rebased to solve merge conflicts in requirements.in / requirements.txt.

File: bazel/external/requirements.in
Comment (jheidbrink) [inline bazel/external/requirements.in]: What is the reason for pinning most of the versions in `requirements.in`? As far as I understand the requirements.in, it is supposed to contain version constraints, and then `pip-compile` generates the `requirements.txt` which has exact pinnings. The `requirements.txt` will then be used by `pip` to install the exactly pinned versions.

This line for example would previously tell `pip-compile` to not upgrade to a version `>= 4.0`. With the line changed to `h2==3.2.0`, `pip-compile` no longer knows about this constraint.

Without the exact pinning, I can use other package repositories like nixpkgs to install Python dependencies. This stops working when we exactly pin the dependencies, because that very specific combination of packages usually isn't available there.

Comment (jheidbrink) [reply bazel/external/requirements.in]: Related question: Why do we convert some of the `>=` lines to `==`, while others stay `>=` constraints?

File: lte/gateway/python/setup.py
Comment (nstng) [inline lte/gateway/python/setup.py]: Is this comment outdated? To me it looks like grpcio-tools>=1.46.3 is used here and grpcio>=1.47.0 in orc8r/setup.py

File: lte/gateway/python/setup.py
Comment (nstng) [inline lte/gateway/python/setup.py]: This seems outdated - I do not see grpc setup in lte/gateway/python/Makefile

File: bazel/external/requirements.in
Comment (jheidbrink) [inline bazel/external/requirements.in]: What is the reason for exactly pinning most packages in `requirements.in`? As far as I understand the mechanism, `pip-compile` will already create a `requirements.txt` with exact pinnings, and pip will use `requirements.txt` to install exactly pinned packages. The `requirements.in` can then be used to define versions constraints for `pip-compile`.

In this example, the previous version of the line would tell `pip-compile` not to upgrade to `h2 >= 4.0`. With the new line `h2==3.2.0`, `pip-compile` no longer has that information.

That scheme also allows different package managers that don't necessarily have all the packages from pypi. For example a developer could choose to install Python dependencies from nixpkgs on their machine. With the exact pinnings in `requirements.in` this doesn't work anymore as nixpkgs won't have that exact combination of packages available.

Comment (sebathomas) [reply bazel/external/requirements.in]: I some cases (e.g. jinja2), things break if you allow `>=` and pip-compile won't be able to resolve it. Not sure about h2 in particular.

Comment (jheidbrink) [reply bazel/external/requirements.in]: Sorry, duplicate comment - for some reason Github didn't show my previous comment and I thought it was gone...

Comment (jheidbrink) [reply bazel/external/requirements.in]: Ah, i see, thanks!

File: lte/gateway/python/setup.py
Comment (sebathomas) [reply lte/gateway/python/setup.py]: Was there a reason to use different versions? I assume we know that it works due to the tests but I would still prefer to keep them the same.

File: lte/gateway/python/setup.py
Comment (MoritzThomasHuebner) [reply lte/gateway/python/setup.py]: I removed the comment

File: lte/gateway/python/setup.py
Comment (MoritzThomasHuebner) [reply lte/gateway/python/setup.py]: I made the versions consistent

File: bazel/external/requirements.in
Comment (MoritzThomasHuebner) [reply bazel/external/requirements.in]: Thanks, I am just copying below what I sent privately on slack:

"I had another look into the [requirements.in](http://requirements.in/) file and how we compile this to a requirements.txt. One thing I have noticed today is that the output of pip-compile depends on what is already in the requirements.txt. In fact, if you delete the requirements.txt on the current state of master and try to build it again, it will fail (see screenshot). To resolve the version conflict in the screenshot, I manually set flask<2.1.0.  pip-compile apparently only changes versions in the requirements.txt if it is forced to. If the existing versions fulfil the [requirements.in](http://requirements.in/), pip-compile will not upgrade versions.
With all of that in mind, I changed the PR so that we only use fixed versions where they were already in place. I incremented all of the fixed versions, added the constraint on flask, and copied the sentry-sdk constraint from the setup.py file. Additionally, I changed systemd to systemd-python. These are actually different packages with similar interfaces. systemd-python is what is actually required in magma.
Finally, I removed the requirements.txt , built it again from scratch, and added the changes to the PR."

**Comment by MoritzThomasHuebner** [comment]
I triggered a run for the lte integ test
https://github.com/MoritzThomasHuebner/magma/actions/runs/2917075867

File: bazel/external/requirements.in
Comment (jheidbrink) [reply bazel/external/requirements.in]: Yes, that seems to be in accordance with https://github.com/jazzband/pip-tools#example-usage-for-pip-compile:
> Note: If pip-compile finds an existing requirements.txt file that fulfils the dependencies then no changes will be made, even if updates are available. To compile from scratch, first delete the existing requirements.txt file, or see [Updating requirements](https://github.com/jazzband/pip-tools#updating-requirements) for alternative approaches.

Comment (alexzurbonsen) [reply bazel/external/requirements.in]: > What is the reason for pinning most of the versions in `requirements.in`? 

That originated from a conversation between @LKreutzer and me, who brought https://github.com/magma/magma/issues/11743 to my attention. The second bullet point, according to our interpretation, implied that one wants to make sure, that same versions are used independently of whether you rely on setup.py or requirements.in.

But just as an explanation, I guess its fine without that as well.

Comment (jheidbrink) [reply bazel/external/requirements.in]: Good point, ensuring that installed versions are the same when using `pip install` vs building & installing the `.deb` package is a good argument for pinning in `requirements.in`.

Comment (jheidbrink) [reply bazel/external/requirements.in]: But I think the current change where we keep the pinning for packages that were already pinned is fine for now.

**Comment by jheidbrink** [comment]
The changes look good to me. I verified i could build all `py_binary`s with Bazel, and run the `subscriberdb` binary which uses the new `systemd-python` package. I did not double-check that I can `fab package` as that was already done. Did someone test the new `.deb` on a freshly destroyed VM so we can be sure that the .deb didn't just work because of some things still laying around in the system?

**Comment by MoritzThomasHuebner** [comment]
> The changes look good to me. I verified i could build all `py_binary`s with Bazel, and run the `subscriberdb` binary which uses the new `systemd-python` package. I did not double-check that I can `fab package` as that was already done. Did someone test the new `.deb` on a freshly destroyed VM so we can be sure that the .deb didn't just work because of some things still laying around in the system?

@jheidbrink yes we tested this in the process and installed magma just from the debian packages.

**Comment by sebathomas** [comment]
I think the integration test run is mostly looking good, only one red test from what I can see. Is test_tau_periodic_active known to be flaky?

**Comment by jheidbrink** [comment]
> Is test_tau_periodic_active known to be flaky?

At least I remember that I commented that one out locally a while ago.

**Review by nstng — APPROVED** [review]
bazel + pipelined lgtm

**Review by jheidbrink — APPROVED** [review]

**Comment by sebathomas** [comment]
Needs another rebase because of changes to the required jobs, then it should finally be ready to merge.

**Pull request merged**
